cmd/tvx/state.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		return nil, err
+	}
+
+	for _, a := range pluck {
+		actor, err := src.GetActor(a)
+		if err != nil {
+			return nil, fmt.Errorf("get actor %s failed: %w", a, err)
+		}
+
+		err = dst.SetActor(a, actor)
+		if err != nil {
+			return nil, err
+		}
+
+		// recursive copy of the actor state.
+		err = vm.Copy(context.TODO(), sg.stores.Blockstore, sg.stores.Blockstore, actor.Head)
+		if err != nil {
+			return nil, err
+		}
+
+		actorState, err := sg.api.ChainReadObj(sg.ctx, actor.Head)
+		if err != nil {
+			return nil, err
+		}
+
+		cid, err := sg.stores.CBORStore.Put(sg.ctx, &cbg.Deferred{Raw: actorState})
+		if err != nil {
+			return nil, err
+		}
+
+		if cid != actor.Head {
+			panic("mismatched cids")
+		}
+	}
+
+	return dst, nil
+}
+
+// saveInitActor saves the state of the init actor to the provided state map.
+func (sg *StateSurgeon) saveInitActor(initActor *types.Actor, initState init_.State, st *state.StateTree) error {
+	log.Printf("saving init actor into state tree")
+
+	// Store the state of the init actor.
+	cid, err := sg.stores.CBORStore.Put(sg.ctx, initState)
+	if err != nil {
+		return err
+	}
+	actor := *initActor
+	actor.Head = cid
+
+	err = st.SetActor(init_.Address, &actor)
+	if err != nil {
+		return err
+	}
+
+	cid, _ = st.Flush(sg.ctx)
+	log.Printf("saved init actor into state tree; new root: %s", cid)
+	return nil
+}
+
+// retainInitEntries takes an old init actor state, and retains only the
+// entries in the retain set, returning a new init actor state.
+func (sg *StateSurgeon) retainInitEntries(state init_.State, retain []address.Address) error {
+	log.Printf("retaining init actor entries for addresses: %v", retain)
+
+	m := make(map[address.Address]struct{}, len(retain))
+	for _, a := range retain {
+		m[a] = struct{}{}
+	}
+
+	var remove []address.Address
+	_ = state.ForEachActor(func(id abi.ActorID, address address.Address) error {
+		if _, ok := m[address]; !ok {
+			remove = append(remove, address)
+		}
+		return nil
+	})
+
+	err := state.Remove(remove...)
+	log.Printf("new init actor state: %+v", state)
+	return err
+}
+
+// resolveAddresses resolved the requested addresses from the provided
+// InitActor state, returning a slice of length len(orig), where each index
+// contains the resolved address.
+func (sg *StateSurgeon) resolveAddresses(orig []address.Address, ist init_.State) (ret []address.Address, err error) {
+	log.Printf("resolving addresses: %v", orig)
+
+	ret = make([]address.Address, len(orig))
+	for i, addr := range orig {
+		resolved, found, err := ist.ResolveAddress(addr)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			return nil, fmt.Errorf("address not found: %s", addr)
+		}
+		ret[i] = resolved
+	}
+
+	log.Printf("resolved addresses: %v", ret)
+	return ret, nil
+}
+
+// loadInitActor loads the init actor state from a given tipset.
+func (sg *StateSurgeon) loadInitActor(st *state.StateTree) (*types.Actor, init_.State, error) {
```

Comment:
will this need to be interface'd to support versioning?

Reply:
Maybe, but honestly I don't think we'll end up using the `accessed-actors` ever. I just don't want to throw it away too quickly in case it comes in handy, but it's very inefficient and ultimately useless.